### PR TITLE
Add a benchmark/profiler to count cycles in demo-prover

### PR DESCRIPTION
**Pull request by dubbelosix**
# Description

We have benchmarking for examples/demo-rollup where we count the number of transactions per second (TPS) by feeding in random data. For demo-prover however, we're interesting in counting the number of cycles that processing a block/transaction takes.
The PR includes
* a cycle_tracker macro which can be used to annotate a function, and wraps it to calculate the cycles the function takes
* io_callback system call and handler to get metrics from guest to host and aggregate them for calculation
* feature gating with "bench" feature to prevent adding un-necessary dependencies for regular run
* build.rs modifications to pass the "bench" feature flag into the guest env

## Docs
sovereign-sdk/examples/demo-prover/host/benches/README.md

### CODE REVIEW

File: adapters/risc0/src/host.rs [hunk 1]
```diff
@@ -16,8 +20,18 @@ pub struct Risc0Host<'a> {
 
 impl<'a> Risc0Host<'a> {
     pub fn new(elf: &'a [u8]) -> Self {
+        #[cfg(not(feature = "bench"))]
```
Comment (bkolad) [inline adapters/risc0/src/host.rs]: We should have another method `new_for_bench` and ` #[cfg(not(feature = "bench"))]` the whole method

File: examples/demo-prover/host/benches/prover_bench.rs [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+use sov_rollup_interface::stf::StateTransitionFunction;
+use sov_rollup_interface::zk::ZkvmHost;
+
+#[cfg(feature = "bench")]
+use risc0_adapter::metrics::GLOBAL_HASHMAP;
+
+use std::collections::HashMap;
+
+#[derive(Debug, Clone, PartialEq, Deserialize)]
+pub struct RollupConfig {
```
Comment (bkolad) [inline examples/demo-prover/host/benches/prover_bench.rs]: We already have this structure in DemoRollup

File: examples/demo-prover/methods/build.rs [hunk 3]
```diff
@@ -1,3 +1,14 @@
+use std::collections::HashMap;
+
 fn main() {
-    risc0_build::embed_methods();
+    #[cfg(not(feature = "bench"))]
```
Comment (bkolad) [inline examples/demo-prover/methods/build.rs]: let's separate all the functionality hidden by #[cfg(feature = "bench")] in a separate method and feature guard this method.

File: examples/demo-prover/methods/guest/src/bin/rollup.rs [hunk 4]
```diff
@@ -69,12 +74,32 @@ pub fn main() {
...
         final_state_root: result.state_root.0,
         validity_condition,
     };
     env::commit(&output);
+
     env::write(&"new state root committed\n");
+
+    let end_cycles = env::get_cycle_count();
+
+    #[cfg(feature = "bench")]
```
Comment (bkolad) [inline examples/demo-prover/methods/guest/src/bin/rollup.rs:95]: Let's make it a operate method

File: zk-cycle-utils/Cargo.toml [hunk 5]
```diff
@@ -0,0 +1,27 @@
+[package]
```
Comment (bkolad) [inline zk-cycle-utils/Cargo.toml]: Let's add README to the crate.

File: zk-cycle-utils/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,54 @@
+extern crate proc_macro;
```
Comment (bkolad) [inline zk-cycle-utils/src/lib.rs]: Let's add
```
#![deny(missing_docs)]
#![doc = include_str!("../README.md")]
```

File: zk-cycle-utils/Cargo.toml [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+name = "zk-cycle-utils"
```
Comment (bkolad) [inline zk-cycle-utils/Cargo.toml]: I think we should have a high level folder `utils` and inside a crate `zk-utils`. The way we have a place where we can to add common functionality. @preston-evans98 WDT?

File: zk-cycle-utils/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+
+use proc_macro::TokenStream;
+use quote::quote;
+use syn::{parse_macro_input, ItemFn};
+
+#[proc_macro_attribute]
+pub fn cycle_tracker(_attr: TokenStream, item: TokenStream) -> TokenStream {
```
Comment (bkolad) [inline zk-cycle-utils/src/lib.rs]: Let's write a tests for this macro?

File: zk-cycle-utils/Cargo.toml (hunk 7 above)
Comment (preston-evans98) [reply zk-cycle-utils/Cargo.toml]: Sounds good to me!

File: examples/demo-prover/host/benches/prover_bench.rs [hunk 9]
```diff
@@ -96,12 +89,12 @@ use std::collections::HashMap;
 #[cfg(feature = "bench")]
 use risc0_adapter::metrics::GLOBAL_HASHMAP;
 
-#[derive(Debug, Clone, PartialEq, Deserialize)]
-pub struct RollupConfig {
-    pub start_height: u64,
-    pub da: DaServiceConfig,
-    pub runner: RunnerConfig,
-}
+// #[derive(Debug, Clone, PartialEq, Deserialize)]
```
Comment (bkolad) [inline examples/demo-prover/host/benches/prover_bench.rs]: Should this be removed?

**Review by bkolad — APPROVED** [review]

**Pull request merged**